### PR TITLE
[AutoRound] Support AutoRound Format Block-Wise FP8 in vLLM

### PR DESCRIPTION
**Pull request by Zhenzhong1**
Test:

```bash
CUDA_VISIBLE_DEVICES=1 python examples/basic/offline_inference/generate.py --model ../tmp_autoround/Llama-3.1-8B-fp-w8g128x128 --enforce-eager --gpu_memory_utilization 0.8

--------------------------------------------------
Prompt: 'Hello, my name is'
Generated text: ' Adelina. I am a Licensed Clinical Social Worker and Certified Clinical Trauma'
--------------------------------------------------
Prompt: 'The president of the United States is'
Generated text: ' the head of state and head of government of the United States. The president directs'
--------------------------------------------------
Prompt: 'The capital of France is'
Generated text: ' Paris, which is also the most populated city in the country. France is the'
--------------------------------------------------
Prompt: 'The future of AI is'
Generated text: ' a topic that has been discussed for many years. The potential benefits of AI are'
--------------------------------------------------

```
```bash
pytest -v \
  tests/quantization/test_auto_round.py::test_inc_config_accepts_block_fp8 \
  tests/quantization/test_auto_round.py::test_inc_config_rejects_invalid_block_fp8_activation_config \
  tests/quantization/test_auto_round.py::test_inc_layer_config_fp8_block_helper \
  tests/quantization/test_auto_round.py::test_inc_resolve_scheme_selects_fp8_block \
  tests/quantization/test_auto_round.py::test_inc_fp8_linear_scheme_delegates_to_fp8_linear_method

tests/quantization/test_auto_round.py::test_inc_config_accepts_block_fp8 PASSED                                                                                                                                                                                         [ 20%]
tests/quantization/test_auto_round.py::test_inc_config_rejects_invalid_block_fp8_activation_config PASSED                                                                                                                                                               [ 40%]
tests/quantization/test_auto_round.py::test_inc_layer_config_fp8_block_helper PASSED                                                                                                                                                                                    [ 60%]
tests/quantization/test_auto_round.py::test_inc_resolve_scheme_selects_fp8_block PASSED                                                                                                                                                                                 [ 80%]
tests/quantization/test_auto_round.py::test_inc_fp8_linear_scheme_delegates_to_fp8_linear_method PASSED     

```

### CODE REVIEW

**Comment by Zhenzhong1** [comment]
@claude review

File: tests/quantization/test_auto_round.py [hunk 1]
```diff
@@ -329,6 +402,89 @@ def test_inc_linear_method_delegates() -> None:
...
+    assert captured["quant_config"].weight_block_size == [128, 128]
+    assert captured["method"].marlin_input_dtype == "fp8-marlin"
+    assert result == "applied"
+    assert [call[0] for call in captured["method"].calls] == [
+        "create_weights",
+        "process_weights_after_loading",
+        "apply",
+    ]
+
+
```
Comment (yiliu30) [inline tests/quantization/test_auto_round.py:1088]: Please extend model test to include such case.

File: vllm/model_executor/layers/quantization/inc/schemes/inc_fp8_scheme.py [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    import torch
+
+    from ..config_parser import INCLayerConfig
+    from ..inc import INCConfig
+
+
+class INCFp8Scheme(INCScheme):
+    @staticmethod
+    def can_handle(layer_config: "INCLayerConfig") -> bool:
+        return layer_config.is_fp8_block
```
Comment (yiliu30) [inline vllm/model_executor/layers/quantization/inc/schemes/inc_fp8_scheme.py:19]: Do we currently only support block-wise granularity? If so, please add a TODO here for adding support for other granularities.

File: vllm/model_executor/layers/quantization/inc/schemes/inc_fp8_scheme.py [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+    def get_linear_method(
+        self,
+        config: "INCConfig",
+        layer: "torch.nn.Module",
+        prefix: str,
+        layer_config: "INCLayerConfig",
+    ):
+        del config, layer
+
+        assert isinstance(layer_config.group_size, tuple)
```
Comment (yiliu30) [inline vllm/model_executor/layers/quantization/inc/schemes/inc_fp8_scheme.py:30]: Please raise error info for other cases.

File: vllm/model_executor/layers/quantization/inc/schemes/inc_wna16_linear.py [hunk 4]
```diff
@@ -70,6 +70,12 @@ def _build_inner_method(self):
         )
 
     def _build_gptq_method(self):
+        group_size = self.layer_config.group_size
```
Comment (yiliu30) [inline vllm/model_executor/layers/quantization/inc/schemes/inc_wna16_linear.py]: Why we need to change `wna16` here?

File: vllm/model_executor/layers/quantization/inc/inc.py [hunk 5]
```diff
@@ -185,8 +309,11 @@ def get_quant_method(self, layer: torch.nn.Module, prefix: str):
...
-        """Override the `auto-round` method to `inc`."""
-        is_auto_round_format = hf_quant_cfg.get("quant_method", None) == "auto-round"
-        if is_auto_round_format:
+        del user_quant, hf_config
+
+        quant_method = hf_quant_cfg.get("quant_method", None)
+        if quant_method in cls.AUTO_ROUND_QUANT_METHODS:
+            return cls.get_name()
+        if quant_method == cls.FP8_BLOCK_PACKING_FORMAT:
             return cls.get_name()
```
Comment (yiliu30) [inline vllm/model_executor/layers/quantization/inc/inc.py:366]: Why we need to change this?

**Review by yiliu30 — CHANGES_REQUESTED** [review]

File: vllm/model_executor/layers/quantization/inc/schemes/inc_wna16_linear.py (hunk 4 above)
Comment (Zhenzhong1) [reply vllm/model_executor/layers/quantization/inc/schemes/inc_wna16_linear.py]: The group_size type is tuple for block-wise model. 
I changed the group_size type here.
class INCLayerConfig:
    bits: int
    group_size: int | tuple[int, int]

so i got this precommit issues.
Image

File: tests/quantization/test_auto_round.py (hunk 1 above)
Comment (Zhenzhong1) [reply tests/quantization/test_auto_round.py:1088]: added

File: vllm/model_executor/layers/quantization/inc/schemes/inc_fp8_scheme.py (hunk 2 above)
Comment (Zhenzhong1) [reply vllm/model_executor/layers/quantization/inc/schemes/inc_fp8_scheme.py:19]: yes. Added.

File: vllm/model_executor/layers/quantization/inc/schemes/inc_fp8_scheme.py (hunk 3 above)
Comment (Zhenzhong1) [reply vllm/model_executor/layers/quantization/inc/schemes/inc_fp8_scheme.py:30]: Fixed.

File: vllm/model_executor/layers/quantization/inc/inc.py (hunk 5 above)
Comment (Zhenzhong1) [reply vllm/model_executor/layers/quantization/inc/inc.py:366]: if don't change this, we will get this error.
image

block-wise hf_cfg: 
```bash
hf_quant_cfg:{'act_bits': 8, 'act_data_type': 'fp', 'act_dynamic': True, 'act_group_size': 128, 'act_sym': True, 'activation_scheme': 'dynamic', 'autoround_version': '0.13.0', 'bits': 8, 'block_name_to_quantize': 'model.layers', 'data_type': 'fp', 'enable_quanted_input': False, 'fmt': 'e4m3', 'group_size': [128, 128], 'quant_method': 'auto_round:fp8', 'sym': True, 'weight_block_size': [128, 128]}
```

For block-wise fp8 model config,  "quant_method": "auto_round:fp8",
other models are  "quant_method": "auto-round",

This change is needed because vLLM validates the top-level quant_method before it parses the INC-specific config. AutoRound block-wise FP8 checkpoints export quant_method as auto_round:fp8, while the existing INC override only recognized auto-round. Without normalizing auto_round:fp8 here, vLLM treats it as an unknown quantization method and fails before reaching the FP8 config logic. MXFP8 does not need this because it still uses quant_method: auto-round.

**Comment by Zhenzhong1** [comment]
@xinyu-intel @jikunshang Hi, please take a reivew when you have time.

**Comment by jikunshang** [comment]
@claude review

File: tests/quantization/test_auto_round.py [hunk 6]
```diff
@@ -73,6 +77,14 @@
...
         id="auto_round:llm_compressor_mxfp8",
     ),
+    pytest.param(
+        "INC4AI/Llama-3.1-8B-fp-w8g128x128-for-ut",
+        marks=pytest.mark.skipif(
+            not (current_platform.is_cuda() or current_platform.is_xpu()),
+            reason="Block-wise AutoRound model only supports CUDA backend for now.",
+        ),
+        id="auto_round:block_wise_fp8_on_cuda",
+    ),
```
[the remark this answers is not in the document]
Comment (Zhenzhong1) [reply tests/quantization/test_auto_round.py:91]: fixed.

File: tests/quantization/test_auto_round.py [hunk 7]
```diff
@@ -450,6 +489,52 @@ def test_inc_layer_config_mx_fp_helpers() -> None:
...
+    assert config.weight_bits == 8
+    assert config.group_size == (128, 128)
+    assert config.sym is True
+    assert config.data_type == "fp"
+    assert config.packing_format == "auto_round:fp8"
+
+
+def test_inc_config_rejects_invalid_block_fp8_activation_config() -> None:
+    with pytest.raises(ValueError, match="act_dynamic=True"):
+        INCConfig.from_config(make_block_fp8_raw_config(act_dynamic=False))
```
[the remark this answers is not in the document]
Comment (Zhenzhong1) [reply tests/quantization/test_auto_round.py:506]: fixed.

File: vllm/model_executor/layers/quantization/inc/inc.py [hunk 8]
```diff
@@ -126,47 +145,120 @@ def is_mxfp8(self) -> bool:
...
+            }
+            for field_name, expected_value in expected_fields.items():
+                actual_value = self.get_from_keys_or(
+                    config, [field_name], expected_value
+                )
+                assert actual_value == expected_value, (
+                    "INC MXFP8 only supports "
+                    f"{field_name}={expected_value!r}, "
+                    f"but found {field_name}={actual_value!r}."
+                )
```
[the remark this answers is not in the document]
Comment (Zhenzhong1) [reply vllm/model_executor/layers/quantization/inc/inc.py]: rejected.

File: tests/quantization/test_auto_round.py [hunk 9]
```diff
@@ -116,6 +128,11 @@
         "gpu_memory_utilization": 0.8,
         "max_model_len": 512,
     },
+    "INC4AI/Qwen3-8B-fp-w8g128x128-for-ut": {
+        "gpu_memory_utilization": 0.8,
+        "enforce_eager": True,
```
Comment (yiliu30) [inline tests/quantization/test_auto_round.py]: Please remove it.

File: vllm/model_executor/layers/quantization/inc/inc.py (hunk 5 above)
Comment (yiliu30) [reply vllm/model_executor/layers/quantization/inc/inc.py:366]: For models generated by AutoRound, `quant_method` should always be set to `auto-round`. This looks like an issue on the AutoRound side. Could you please fix it there? Thanks!

**Review by yiliu30 — CHANGES_REQUESTED** [review]

File: tests/quantization/test_auto_round.py (hunk 9 above)
Comment (Zhenzhong1) [reply tests/quantization/test_auto_round.py]: removed.

File: vllm/model_executor/layers/quantization/inc/inc.py (hunk 5 above)
Comment (Zhenzhong1) [reply vllm/model_executor/layers/quantization/inc/inc.py:366]: WIP: https://github.com/intel/auto-round/pull/2138

File: vllm/model_executor/layers/quantization/inc/inc.py [hunk 10]
```diff
@@ -49,24 +52,40 @@ class INCConfig(QuantizationConfig):
...
+    @staticmethod
+    def _normalize_data_type(data_type: str) -> str:
+        # Keep a small alias for branches that already experimented with "fp8".
+        return "fp" if data_type == "fp8" else data_type
+
     def __init__(
         self,
         weight_bits: int,
-        group_size: int,
+        group_size: int | list[int] | tuple[int, int],
```
Comment (jikunshang) [inline vllm/model_executor/layers/quantization/inc/inc.py]: should we reuse group_size for int/mxfpX/block FP8? if we want to unify them, we'd better always use `list[int]`. if it's never be shared, should we use group_size and block_size seprately?

File: vllm/model_executor/layers/quantization/inc/inc.py [hunk 11]
```diff
@@ -126,47 +145,120 @@ def is_mxfp8(self) -> bool:
...
+                f"but found group_size={self.group_size!r}."
+            )
+            assert self.packing_format in int_packing_formats, (
+                "INC int quantization only supports "
+                f"packing_format in {sorted(int_packing_formats)!r}, "
+                f"but found packing_format={self.packing_format!r}."
+            )
+
+        elif self.data_type == self.FP8_BLOCK_DATA_TYPE:
+            assert self.weight_bits == self.FP8_BLOCK_BITS, (
```
Comment (jikunshang) [inline vllm/model_executor/layers/quantization/inc/inc.py]: I feel we don't need so many assert..

File: vllm/model_executor/layers/quantization/inc/schemes/factory.py [hunk 12]
```diff
@@ -9,11 +9,13 @@
 
 
 def resolve_scheme(layer_config: "INCLayerConfig") -> "INCScheme":
+    from .inc_fp8_scheme import INCFp8Scheme
```
Comment (jikunshang) [inline vllm/model_executor/layers/quantization/inc/schemes/factory.py:12]: it's block-fp8 only? what about per-tensor, per-channel fp8?

File: vllm/model_executor/layers/quantization/inc/inc.py (hunk 5 above)
Comment (Zhenzhong1) [reply vllm/model_executor/layers/quantization/inc/inc.py:366]: Hi, it's merged.

**Review by yiliu30 — APPROVED** [review]
LGTM

File: vllm/model_executor/layers/quantization/inc/inc.py (hunk 11 above)
Comment (Zhenzhong1) [reply vllm/model_executor/layers/quantization/inc/inc.py]: It's because there are many different date types here. I refactored this part. [335d92f](https://github.com/vllm-project/vllm/pull/47434/commits/335d92fa7f8d226224a997a243b467848659b6c3)

File: vllm/model_executor/layers/quantization/inc/inc.py (hunk 10 above)
Comment (Zhenzhong1) [reply vllm/model_executor/layers/quantization/inc/inc.py]: We can reuse group_size for int/mxfpX/block FP8. For INT and MXFP, the value is naturally a scalar. For block-wise FP8, it is naturally a 2-D shape. They do share the same high-level meaning, namely the granularity of weight quantization, so I do not see a strong need to split the external configuration into `group_size` and `block_size`.

On the other hand, standardizing all cases as `list[int]` would not really simplify the model. We would still need additional validation, similar to `_normalize_group_size` now in this PR, to distinguish scalar group-wise cases from 2-D block-wise cases, while also introducing extra and unnecessary changes in model generation.

File: vllm/model_executor/layers/quantization/inc/schemes/factory.py (hunk 12 above)
Comment (Zhenzhong1) [reply vllm/model_executor/layers/quantization/inc/schemes/factory.py:12]: This is the first vervsion for FP8 scheme.  So this PR covers FP8 BLOCK only. Per-tensor and per-channel FP8 will be added once ARK FP8 is ready.

File: vllm/model_executor/layers/quantization/inc/inc.py (hunk 10 above)
Comment (Zhenzhong1) [reply vllm/model_executor/layers/quantization/inc/inc.py]: Hi, `group_size` can be either an `int` or a `tuple`. It's ok I think. no list any more.
[9c0688f](https://github.com/vllm-project/vllm/pull/47434/commits/9c0688f0ef05bcc048bf50dc933a7bd1f1608269)

File: vllm/model_executor/layers/quantization/inc/schemes/inc_fp8_linear.py [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+
+
+class INCFp8LinearScheme(INCLinearScheme):
+    def __init__(self, prefix: str, weight_block_size: tuple[int, int]) -> None:
+        self.quant_config = Fp8Config(
+            is_checkpoint_fp8_serialized=True,
+            activation_scheme="dynamic",
+            weight_block_size=list(weight_block_size),
+        )
+        self.inner_method = Fp8LinearMethod(self.quant_config)
```
Comment (jikunshang) [inline vllm/model_executor/layers/quantization/inc/schemes/inc_fp8_linear.py]: inner_method -> linear_method ?

Comment (Zhenzhong1) [reply vllm/model_executor/layers/quantization/inc/schemes/inc_fp8_linear.py]: I can do this, but many files all use 'inner_method' as the linear method name. For consistency with other files, the name 'inner_method' is also acceptable overall ？

Comment (jikunshang) [reply vllm/model_executor/layers/quantization/inc/schemes/inc_fp8_linear.py]: most are using `quant_method`.

Comment (Zhenzhong1) [reply vllm/model_executor/layers/quantization/inc/schemes/inc_fp8_linear.py]: fixed. 
032d126

**Review by jikunshang — APPROVED** [review]
/ci run

**Comment by Zhenzhong1** [comment]
@jikunshang INTEL-CI passed. could u pls trigger the CI again? seems it's still pending?

**Comment by jikunshang** [comment]
you can trigger by comment `/ci run` as well.
